primer-miso/Makefile.wasm32

```diff
@@ -0,0 +1,26 @@
+# NOTE:
+#
+# This Makefile assumes you're using the `nix develop .#wasm` shell.
+
+CABAL = wasm32-wasi-cabal
+
+build:	configure
+	$(CABAL) build
+
+frontend: build
+	./build-frontend.sh
+
+frontend-prod: build
+	./build-frontend.sh -Oz
+
+serve-frontend: frontend
+	simple-http-server -i -p 8000 -- dist
```

Comment:
Maybe we should add some instructions to the README for this at least? It wasn't immediately obvious that the command to run the production server is `nix develop .#wasm -c make -f Makefile.wasm32 serve-frontend`.

Reply:
Alternatively we could presumably create an actual Nix output for this? Note that in the absence of Wasm support from Haskell.nix, we just get a `nix run .#primer-miso:exe:primer-miso` that doesn't actually do anything. This is because when we compile natively, we currently use a `main` function which only really works in GHCI...

Reply:
We could alter our use of Jsaddle to give us a usable native binary, but honestly I'm not sure it's useful.

What might be useful is moving the comment from `main` in to an actual make target for ease of development. And potentially substituting GHCId with [ghciwatch](https://github.com/mercurytechnologies/ghciwatch), which is meant to be a modern replacement.

Reply:
> What might be useful is moving the comment from `main` in to an actual make target for ease of development.

_Also_, the command needs modifying at least slightly due to the project structure. We now require `ghcid -c "cabal repl exe:primer-miso" -W -T ':main'`. This still isn't very useful though, since it only picks up changes to `Main.hs`...

What we really want is `ghcid -c "cabal repl --enable-multi-repl exe:primer-miso lib:primer-miso" -W -T 'Main.main'`, or `ghcid -c "cabal repl --enable-multi-repl exe:primer-miso all" -W -T 'Main.main'`. This should work according to [here](https://www.well-typed.co.uk/blog/2023/03/cabal-multi-unit/) and [here](https://github.com/ndmitchell/ghcid/issues/208), and even does for [my minimal example project](https://github.com/georgefst/hello-hs/compare/master...multi-component). But for whatever reason, I find that file changes are never picked up, i.e. when I save a file GHCId doesn't actually do anything.

@dhess I'd be interested to see whether you can reproduce this. It's a pretty crucial part of the developer experience.

(FWIW, Ghciwatch doesn't support this yet, but it's on the radar: https://github.com/MercuryTechnologies/ghciwatch/issues/316.)

Reply:
> Alternatively we could presumably create an actual Nix output for this? Note that in the absence of Wasm support from Haskell.nix, we just get a `nix run .#primer-miso:exe:primer-miso` that doesn't actually do anything. This is because when we compile natively, we currently use a `main` function which only really works in GHCI...

I don't think so, because the Wasm compilers we currently get via `ghc-meta-wasm` aren't hooked into either `haskell.nix` or nixpkgs's `haskellPackages`, so there's currently no way to build the dependencies and store them in the Nix store, as would be required for a proper Nix derivation.

Reply:
> We could alter our use of Jsaddle to give us a usable native binary, but honestly I'm not sure it's useful.
> 
> What might be useful is moving the comment from `main` in to an actual make target for ease of development. And potentially substituting GHCId with [ghciwatch](https://github.com/mercurytechnologies/ghciwatch), which is meant to be a modern replacement.

Using `make` for the dev target sounds like a good idea, so I'll add that.

Even though the native executable isn't useful, I still want to build it (via Nix) just to catch any issues in the Nix CI builds.

Reply:
> Maybe we should add some instructions to the README for this at least? It wasn't immediately obvious that the command to serve the compiled Wasm is `nix develop .#wasm -c make -f Makefile.wasm32 serve-frontend`.

Yes, I just haven't gotten around to this yet.

Reply:
I've added some instructions to the `wasm.md` documentation on how to serve the new frontend. Once the new Wasm frontend is good enough to replace the TypeScript version, I'll update all the docs (including the top-level README) to reflect the new development workflows.

Reply:
Is this breakage a consequence of breaking `primer-miso` out into a separate lib and exe? If so, and you think it's critical to the development workflow, I guess we could undo that. (However, I'd probably want to do that as a separate PR after this one, as I'd like there to be some record of why we did it.)

Reply:
> Is this breakage a consequence of splitting primer-miso out into a separate lib and exe?

Yes, but let's keep all this as it is for now, while I continue investigating what can be fixed upstream.

> i.e., using `wasm32-wasi-cabal` rather than native `cabal`

To be honest, I'm surprised it even got close to working with `wasm32-wasi-cabal`. I'm guessing it got further than your native `cabal` attempt because the latter used an older version of Cabal?

> However, I confess I'm not sure what the benefit of this is vs just recompiling, which is pretty fast.

It's the difference between (very roughly) a 10s feedback loop and 0.5s, so it's a gamechanger for experimenting with certain kinds of changes. Admittedly, that 10s is much lower than it would have been with GHCJS, and I don't know whether `jsaddle-warp` would have been developed otherwise. The longer-term plan is some sort of GHCI-in-browser workflow, but I suspect that's still a few releases away.